### PR TITLE
feat: add express session

### LINKED ISSUE #107: Session support to store user credentials
Express session support - so that a multiple requests can be considered a user session

Use cookie support for session; `cookie.sid` contains reference to session, express session deserializes from cookie id

Memory store for session initially

### PR DESCRIPTION
**Pull request by nictownsend**
- in memory session
- cookie keyed by server name

Contributes to: strimzi/strimzi-ui#107

Signed-off-by: Nic Townsend <nictownsend@uk.ibm.com>

### CODE REVIEW

File: server/core/app.ts [hunk 1]
```diff
@@ -26,6 +27,16 @@ export const returnExpress: (
   // add pino-http middleware
   app.use(generateHttpLogger());
 
+  //Add session middleware
+  const sessionOpts: session.SessionOptions = {
+    secret: 'CHANGEME',
+    name: serverName,
+    cookie: {
+      maxAge: 1000 * 3600 * 24 * 30,
```
Comment (JQrdan) [inline server/core/app.ts]: Could you add a comment to say how long this is

File: server/core/app.ts [hunk 2]
```diff
@@ -26,6 +27,16 @@ export const returnExpress: (
   // add pino-http middleware
   app.use(generateHttpLogger());
 
+  //Add session middleware
+  const sessionOpts: session.SessionOptions = {
```
Comment (JQrdan) [inline server/core/app.ts]: Would be neater to import SessionOptions independently

**Review by JQrdan — CHANGES_REQUESTED** [review]
Just a couple of minor things but otherwise LGTM

**Review by pmuir — APPROVED** [review]

File: server/core/app.ts [hunk 3]
```diff
@@ -26,6 +27,16 @@ export const returnExpress: (
   // add pino-http middleware
   app.use(generateHttpLogger());
 
+  //Add session middleware
+  const sessionOpts: SessionOptions = {
```
Comment (matthew-chirgwin) [inline server/core/app.ts:32]: I assume (in later issues) this will be enabled/added only when needed?

File: server/core/core.steps.ts [hunk 4]
```diff
@@ -37,90 +37,65 @@ When(
...
+    };
 
-        Object.entries(headerExpectedValue).forEach(([key, value]) =>
-          expect(headers[key]).toEqual(value)
-        );
-      },
-      (err) => {
-        throw err;
-      }
+    return Object.entries(headerExpectedValue).reduce(
```
Comment (matthew-chirgwin) [inline server/core/core.steps.ts]: This is a lot more subtle/less readable than the prior version IMO. Can a comment be added to explain what this is doing/how the `chainedRequest` is being invoked?

File: server/test_common/commonServerSteps.ts [hunk 5]
```diff
@@ -67,7 +69,22 @@ And(
   stepWhichUpdatesWorld((world) => {
     return {
       ...world,
-      server: request(returnExpress('test-server', () => world.configuration)),
+      server: request(
+        returnExpress(
+          world.serverName || 'test-server',
```
Comment (matthew-chirgwin) [inline server/test_common/commonServerSteps.ts]: WIth the removal of the name argument this will need to be updated

**Review by matthew-chirgwin — CHANGES_REQUESTED** [review]
LGTM - just one test needs a comment for me to (when we return to this in future) we can see how the assertions are being chained

File: server/test_common/commonServerSteps.ts (hunk 5 above)
Comment (nictownsend) [reply server/test_common/commonServerSteps.ts]: Just spotted it :'(

File: server/core/core.steps.ts (hunk 4 above)
Comment (nictownsend) [reply server/core/core.steps.ts]: I'll revert - if it needs a comment then the efficiency isn't worth the readable impact

**Review by scholzj — APPROVED** [review]

**Review by dlabaj — APPROVED** [review]

File: config/static.ts [hunk 6]
```diff
@@ -36,6 +36,9 @@ const server: Config<Literal> = {
         port: 9080,
         transport: {},
       },
+      session: {
```
Comment (matthew-chirgwin) [inline config/static.ts:39]: no issue with the being here - can the typing and doc be updated to reflect? https://github.com/matthew-chirgwin/strimzi-ui/tree/master/server#configuration-options / https://github.com/matthew-chirgwin/strimzi-ui/blob/master/server/types.ts

**Review by matthew-chirgwin — COMMENTED** [review]
LGTM - just one comment re typing/doc

File: server/core/app.ts (hunk 3 above)
Comment (nictownsend) [reply server/core/app.ts:32]: I think it's fair to always have a session as we might want to store information for the lifetime of a browser session. In particular, for logging - @ajborley, would it be useful to store a "client id" in the session?

File: config/static.ts (hunk 6 above)
Comment (nictownsend) [reply config/static.ts:39]: Adding

File: server/types.ts [hunk 7]
```diff
@@ -79,6 +83,7 @@ export type serverConfig = {
   hostname: string;
   /** The port number to use/accept traffic on */
   port: number;
+  session: sessionConfigType;
```
Comment (matthew-chirgwin) [inline server/types.ts:88]: JSDoc for usage/context/purpose?

**Review by matthew-chirgwin — APPROVED** [review]

**Pull request merged**